### PR TITLE
Feature(app): Reactify FCC!

### LINKED ISSUE #4464: reactify challenges

### LINKED ISSUE #5072: Waypoint: Clone an Element Using jQuery: I completed the code and now the phone simulator is showing THREE target5 buttons!
Challenge [Waypoint: Clone an Element Using jQuery](http://www.freecodecamp.com/challenges/waypoint-clone-an-element-using-jquery#?solution=fccss%0A%20%20%24(document).ready(function()%20%7B%0A%20%20%20%20%24(%22%23target1%22).css(%22color%22%2C%20%22red%22)%3B%0A%20%20%20%20%24(%22%23target1%22).prop(%22disabled%22%2C%20true)%3B%0A%20%20%20%20%24(%22%23target4%22).remove()%3B%0A%20%20%20%20%24(%22%23target2%22).appendTo(%22%23right-well%22)%3B%0A%20%20%20%20%24(%22%23target5%22).clone().appendTo(%22%23left-well%22)%3B%0A%20%20%7D)%3B%0Afcces%0A%0A%3C!--%20Only%20change%20code%20above%20this%20line.%20--%3E%0A%0A%3Cdiv%20class%3D%22container-fluid%22%3E%0A%20%20%3Ch3%20class%3D%22text-primary%20text-center%22%3EjQuery%20Playground%3C%2Fh3%3E%0A%20%20%3Cdiv%20class%3D%22row%22%3E%0A%20%20%20%20%3Cdiv%20class%3D%22col-xs-6%22%3E%0A%20%20%20%20%20%20%3Ch4%3E%23left-well%3C%2Fh4%3E%0A%20%20%20%20%20%20%3Cdiv%20class%3D%22well%22%20id%3D%22left-well%22%3E%0A%20%20%20%20%20%20%20%20%3Cbutton%20class%3D%22btn%20btn-default%20target%22%20id%3D%22target1%22%3E%23target1%3C%2Fbutton%3E%0A%20%20%20%20%20%20%20%20%3Cbutton%20class%3D%22btn%20btn-default%20target%22%20id%3D%22target2%22%3E%23target2%3C%2Fbutton%3E%0A%20%20%20%20%20%20%20%20%3Cbutton%20class%3D%22btn%20btn-default%20target%22%20id%3D%22target3%22%3E%23target3%3C%2Fbutton%3E%0A%20%20%20%20%20%20%3C%2Fdiv%3E%0A%20%20%20%20%3C%2Fdiv%3E%0A%20%20%20%20%3Cdiv%20class%3D%22col-xs-6%22%3E%0A%20%20%20%20%20%20%3Ch4%3E%23right-well%3C%2Fh4%3E%0A%20%20%20%20%20%20%3Cdiv%20class%3D%22well%22%20id%3D%22right-well%22%3E%0A%20%20%20%20%20%20%20%20%3Cbutton%20class%3D%22btn%20btn-default%20target%22%20id%3D%22target4%22%3E%23target4%3C%2Fbutton%3E%0A%20%20%20%20%20%20%20%20%3Cbutton%20class%3D%22btn%20btn-default%20target%22%20id%3D%22target5%22%3E%23target5%3C%2Fbutton%3E%0A%20%20%20%20%20%20%20%20%3Cbutton%20class%3D%22btn%20btn-default%20target%22%20id%3D%22target6%22%3E%23target6%3C%2Fbutton%3E%0A%20%20%20%20%20%20%3C%2Fdiv%3E%0A%20%20%20%20%3C%2Fdiv%3E%0A%20%20%3C%2Fdiv%3E%0A%3C%2Fdiv%3E%0A) has an issue.
User Agent is: <code>Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.73 Safari/537.36</code>.
Please describe how to reproduce this issue, and include links to screenshots if possible.

## Issue

I believe I've found bug in the phone simulation in **Waypoint: Clone an Element Using jQuery**:

I entered the code to clone ```target5``` and append it to ```left-well```, and now I see three **target5** buttons in the phone simulator. FCC says my code is correct and advances me to the next challenge. The following challenges also show three target5 buttons:
* Waypoint: Target the Parent of an Element Using jQuery
* Waypoint: Target the Children of an Element Using jQuery

@qualitymanifest confirms this issue on his **Linux** box.

image

## My code:
```html
<script>
  $(document).ready(function() {
    $("#target1").css("color", "red");
    $("#target1").prop("disabled", true);
    $("#target4").remove();
    $("#target2").appendTo("#right-well");
    $("#target5").clone().appendTo("#left-well");
  });
</script>

<!-- Only change code above this line. -->

<div class="container-fluid">
  <h3 class="text-primary text-center">jQuery Playground</h3>
  <div class="row">
    <div class="col-xs-6">
      <h4>#left-well</h4>
      <div class="well" id="left-well">
        <button class="btn btn-default target" id="target1">#target1</button>
        <button class="btn btn-default target" id="target2">#target2</button>
        <button class="btn btn-default target" id="target3">#target3</button>
      </div>
    </div>
    <div class="col-xs-6">
      <h4>#right-well</h4>
      <div class="well" id="right-well">
        <button class="btn btn-default target" id="target4">#target4</button>
        <button class="btn btn-default target" id="target5">#target5</button>
        <button class="btn btn-default target" id="target6">#target6</button>
      </div>
    </div>
  </div>
</div>

```

### LINKED ISSUE #5111: No visual feedback when wrong code is submitted.
The `Run tests (ctrl + enter)` button in waypoints doesn’t give any visual feedback when the wrong code is submitted. 

It would be nice to have this, especially useful to a novice `Camper`.

### LINKED ISSUE #6607: Update /challenges/challenge-name to be /challenges/challenge-block/challenge-name
We want to do this for more specific routes. 
- [ ] make sure old challenges/[partial challenge name] regex retrieval still works
- [ ] make sure map and code portfolio redirects still work
- [ ] We also need to make sure these more specific routes are picked up by the sitemap.

For example, if I type freecodecamp.com/challenges/reverse, this logic should be smart enough to figure out that's /challenges/basic-algorithm-scripting/reverse-a-string and redirect accordingly.

### LINKED ISSUE #6608: Create world language selection options for code portfolio and update user object
They would go down here:

They will need to tie into the user model - something like "worldLanguage": "es".

### LINKED ISSUE #6609: Add world language support
~~Note that we want to make the Spanish versions of challenges available at:~~
/es/challenges/challenge-block/challenge-name

If you:
1. visit the Spanish challenge name URL or 
2. visit an English challenge but have Spanish selected as your "worldLanguage" attribute, that challenge's description and name should be in Spanish.

edit(@berkeleytrue):

All urls will have language encoded within. The default will be `/en`.

* Users will set up their preference in the settings page
* Users will chose from those langauges that are 99% complete

### LINKED ISSUE #6910: Video challenges : Map link redirects to "/map" instead of drawer
In the video challenges, the Map [link on the navbar](http://www.freecodecamp.com/videos/javascript-lingo-variables--camelcase) is redirecting to http://www.freecodecamp.com/map instead of opening the drawer as in other coding challenges.

### LINKED ISSUE #7831: Error on loading a challenge under the /videos path
image
/videos/hike-computer-basics-1-the-4-basic-parts-of-a-computer

Found when running the application on localhost

@BerkeleyTrue

### LINKED ISSUE #8290: Add hints to new react challenge framework

### LINKED ISSUE #9157: Challenges should not auto-run
A simple way to replicate the problem, taking a random challenge:

https://www.freecodecamp.com/challenges/understanding-boolean-values#?solution=alert("BOO!")

This not only pops up an alert, it does it twice.  It shouldn't even be running it at all, unless and until the submit button (or ctrl-enter) requests it.  All the problems people have with infinite loops, and the magic URL parameters people have to enter to prevent them, can be prevented if FCC simply refrains from auto-running user code in the first place.  Infinite loops could then be broken out of by simply hitting refresh.

### PR DESCRIPTION
**Pull request by BerkeleyTrue**
- [x] build javascript/html view
- [x] build step view
- [x] add data fetching
- [x] add test suite rendering from data
- [x] build out challengeFramework
  - [x] convert to CJS
  - [x] run tests in iframe
  - [x] run preview/tests/output in separate frames
  - [x] console logging to output view
- [x] add map view
- [x] ~~video~~ [now called project] challenge view
- [x] Submit/next challenge toast
- [x] move to next challenge
### BETA ready
- [x] move Videos under same folder as other challenges (in progress)
- [x] move back a challenge
- [x] add language routing
- [x] move hints from challenge articles to JSON files
  - [x] define JSON hint structure
  - [x] create react hint elements and redux actions for enabling
- [x] fix map view (drawer broken)
- [x] add LightBox to step challenges
- [x] add map view search/filter
- [x] add hot keys
- [ ] generate code URI on request
- [x] add GA url tracking
### QA (Bring down the hamma!)
- [ ] Fix all da bugs!
### Post merge
- [ ] Modern challenge interface

~~MDN links~~ MDN links will be moved to the hint system
### Changes that will need to made to all challenges
- Challenges that use to implicitly display function output will no longer do so. This challenges will need to use a console log to display that information to the user.

closes #4464
closes #5072
closes #5111
closes #6607
closes #6608
closes #6609
closes #6910 
closes #7831 
closes #8290 
closes #9157

### CODE REVIEW

File: client/sagas/err-saga.js [hunk 1]
```diff
@@ -1,20 +1,14 @@
...
-    const result = next(action);
-    if (!action.error) { return result; }
-
-    console.error(action.error);
-    return dispatch({
+export default function errorSaga(action$) {
+  return action$
+    .filter(({ error }) => !!error)
+    .map(({ error }) => error)
+    .doOnNext(error => console.error(error))
```
Comment (jsJunky) [inline client/sagas/err-saga.js]: :question: Can you also just use `.doOnNext(console.error)` or is that not valid?

Comment (BerkeleyTrue) [inline client/sagas/err-saga.js]: You could, but you would need to bind the method to the console object
```js
.doOnNext(console.error.bind(console))
```

Comment (jsJunky) [inline client/sagas/err-saga.js]: Yeah that looks pretty gross. Why does referencing the console object behave differently when outside of the doOnNext scope? I'll play around with that this weekend Thanks for responding!

Comment (BerkeleyTrue) [inline client/sagas/err-saga.js]: In the doOnNext method call you are passing in a function. If you pass in the method from the console object it loses it's internal context (the `this` object internally becomes the `this` object of the callee function). 

Binding it to the console object ensures that its internal context cannot change.

Comment (jsJunky) [inline client/sagas/err-saga.js]: Ah so it auto binds this much like arrow functions. Thanks!

File: client/frame-runner.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
+document.addEventListener('DOMContentLoaded', function() {
+  var common = parent.__common;
+  var frameId = window.__frameId;
+  var frameReady = common[frameId + 'Ready$'] || { onNext() {} };
+  var Rx = document.Rx;
+  var chai = parent.chai;
+  var source = document.__source;
+
+  document.__getJsOutput = function getJsOutput() {
```
Comment (Grimmreaper81) [inline client/frame-runner.js]: Spread(

**Comment by raisedadead** [comment]
@BerkeleyTrue
###  :sparkles: AWESOME! YOU SIR ARE THE ROCKSTAR :sparkles:

**Comment by atjonathan** [comment]
@BerkeleyTrue :heart:

**Pull request merged**